### PR TITLE
update scoring and add timestamps on guesses

### PR DESCRIPTION
**Pull request by josephmaxwellhyland**

### CODE REVIEW

File: back-end/db/player.ts [hunk 1]
```diff
@@ -141,6 +141,37 @@ export default {
     }
   },
 
+  awardAllPlayersConsolationPoints: async function(guessingPlayers: IPlayer[], quizQIndex: number): Promise<any> {
+    const guessTime = p => p.quizGuesses[quizQIndex].timestamp;
+    const timeBetweenPlayerGuesses = (a: IPlayer, b: IPlayer) => guessTime(b) - guessTime(a);
+
+    guessingPlayers.sort(timeBetweenPlayerGuesses);
```
Comment (josephmaxwellhyland) [inline back-end/db/player.ts:148]: Sort the players by their guess times - whoever answered fastest will be last in the list

File: back-end/db/player.ts [hunk 2]
```diff
@@ -141,6 +141,37 @@ export default {
...
+    for (let i = 0; i < guessingPlayers.length; i++) {
+      const currentPlayer = guessingPlayers[i];
+      await this.awardPlayerConsolationPoints(currentPlayer.id, 25 * (i+1));
```
Comment (josephmaxwellhyland) [inline back-end/db/player.ts:151]: calculate points to award based on array position

File: back-end/db/player.ts [hunk 3]
```diff
@@ -149,7 +180,11 @@ export default {
         throw `Player not found: ${playerId}`;
       } else {    
         const newQuizGuesses = player.quizGuesses;
-        newQuizGuesses[gameData.currentQuestionIndex] = guess;
+        newQuizGuesses[gameData.currentQuestionIndex] = {
```
Comment (josephmaxwellhyland) [inline back-end/db/player.ts:183]: use guess object to store the guess, including the current time

File: back-end/db/player.ts [hunk 4]
```diff
@@ -175,14 +210,9 @@ export default {
...
-  getPlayerGuessesForQuizQuestion: async function(gameId: number, questionIndex: number): Promise<Guess[]> {
+  getPlayerGuessesForQuizQuestion: async function(gameId: number, questionIndex: number): Promise<IGuess[]> {
     const playersInGame = await this.getPlayers(gameId);
-    const playerGuesses = playersInGame.map(p => ({
-      name: p.name,
-      guess: p.quizGuesses[questionIndex]
-    }));
-
-    return playerGuesses;
+    return playersInGame.map(p => p.quizGuesses[questionIndex]);
```
Comment (josephmaxwellhyland) [inline back-end/db/player.ts:215]: grab the guess object right from the db

File: back-end/handlers/hostHelpers.ts [hunk 5]
```diff
@@ -153,46 +153,51 @@ const hostPreAnswer = async (gameId: number, io: Server): Promise<void> => {
...
-  const players = await playerDb.getPlayers(gameId);
-  const guesses = await playerDb.getPlayerGuessesForQuizQuestion(gameId, gameData.currentQuestionIndex);
+  const handsFreeMode = gameData.settings.handsFreeMode;
+  const timePerAnswer = (gameData.settings.timePerAnswer || 10) * 1000;  
+  const currentQuestionIndex = gameData.currentQuestionIndex;
+  const quizLength = gameData.quizQuestions.length || 5;
 
+  const players = await playerDb.getPlayers(gameId);
+  const subjectPlayerId = gameData.quizQuestions[currentQuestionIndex].playerId;
+  const guessingPlayers = players.filter(p => p.id !== subjectPlayerId);
```
Comment (josephmaxwellhyland) [inline back-end/handlers/hostHelpers.ts:167]: list of all players other than the subject

File: back-end/handlers/hostHelpers.ts [hunk 6]
```diff
@@ -153,46 +153,51 @@ const hostPreAnswer = async (gameId: number, io: Server): Promise<void> => {
...
+  const numCorrect = guesses.filter(g => g &&g.guess === correctGuess).length;
 
-  let ScoreAdder = 0;
-  let correctGuess = gameData.quizQuestions[gameData.currentQuestionIndex].correctAnswerIndex
-  for (let i = 0; i < guesses.length; i++) {
-    if (guesses[i].guess === correctGuess) {
-      ScoreAdder += 100;
-    }
+  if (numCorrect === 0) {
+    playerDb.awardAllPlayersConsolationPoints(guessingPlayers, currentQuestionIndex);
```
Comment (josephmaxwellhyland) [inline back-end/handlers/hostHelpers.ts:175]: if no one would have gotten points, all guessers receive some points

File: back-end/handlers/hostHelpers.ts [hunk 7]
```diff
@@ -153,46 +153,51 @@ const hostPreAnswer = async (gameId: number, io: Server): Promise<void> => {
...
+  const newSubjectPlayerState = numCorrect === 0 ? PlayerStates.SeeingAnswerIncorrect : PlayerStates.SeeingAnswerCorrect
+  playerDb.updatePlayerState(subjectPlayer.id, newSubjectPlayerState, io, {});
+
+  for (let i = 0; i < guessingPlayers.length; i++) {
+    const currentPlayer = guessingPlayers[i];
+    const currentPlayerCurrentGuess = currentPlayer.quizGuesses[currentQuestionIndex].guess;
+    const playerCorrect = currentPlayerCurrentGuess === correctGuess;
+    const currentPlayerNewState = playerCorrect ? PlayerStates.SeeingAnswerCorrect : PlayerStates.SeeingAnswerIncorrect;
+
+    await playerDb.updatePlayerState(currentPlayer.id, currentPlayerNewState, io, {});
```
Comment (josephmaxwellhyland) [inline back-end/handlers/hostHelpers.ts:198]: update each player state depending on if they got it correct or not

File: back-end/interfaces/IGuess.ts [hunk 8]
```diff
@@ -0,0 +1,5 @@
+export default interface IGuess {
```
Comment (josephmaxwellhyland) [inline back-end/interfaces/IGuess.ts:1]: new interface for guesses stored in db

File: back-end/interfaces/IPlayer.ts [hunk 9]
```diff
@@ -1,10 +1,11 @@
 import { IPlayerState } from "./IPlayerState"
+import IGuess from "./IGuess"
 
 export default interface IPlayer {
 	name: string,
 	id: string,
 	questionnaireAnswers: string[],
-	quizGuesses: number[],
+	quizGuesses: IGuess[],
```
Comment (josephmaxwellhyland) [inline back-end/interfaces/IPlayer.ts:8]: store list of guess objects

File: front-end/src/host/HostShowAnswer.tsx [hunk 10]
```diff
@@ -78,9 +76,10 @@ export default function HostShowAnswer(props: IShowAnswerProps) {
       res += `! ${encourage} `;
     } else {
       const sad = pickOne([
-        "Unfortunately, no one received any points.",
-        "I'm sorry, no points this round.",
-        "No points this round. Please try to do better next time.",
+        "Unfortunately, no one got it. Luckily, you still get points for failing.",
+        "I'm sorry, this is sad. No one won, but guessers get consolation points for fast answers.",
+        `No points for ${playerName}. Please try to do better next time.`,
```
Comment (josephmaxwellhyland) [inline front-end/src/host/HostShowAnswer.tsx:81]: update announcements for new scoring system

File: front-end/src/host/HostShowAnswer.tsx [hunk 11]
```diff
@@ -106,25 +105,22 @@ export default function HostShowAnswer(props: IShowAnswerProps) {
...
-            }}
-          >
-            {100 *
-              playerGuesses.filter((g) => g.guess === correctAnswerIndex)
-                .length}{" "}
-            points for {playerName}
-          </p>
+          {numCorrect > 0 ?
+            <p style={{margin: 0, fontFamily: `"Concert One", sans-serif`}}>{pctCorrect}% correct = <span style={{fontWeight: "bold", color: "#8080ff"}}>{subjectBonus} points</span> for {playerName}!</p> :
+            <p style={{margin: 0, fontFamily: `"Concert One", sans-serif`}}>0% correct ☹ = consolation points based on speed</p>
```
Comment (josephmaxwellhyland) [inline front-end/src/host/HostShowAnswer.tsx:121]: dynamically show message depending on results

File: front-end/src/player/PlayerApp.tsx [hunk 12]
```diff
@@ -116,10 +119,10 @@ export default function PlayerApp(props: PlayerAppProps) {
       return <PlayerIsSubject />;
     } else if (playerState === "seeing-answer-correct") {
       bottomButtons = false;
-      return <PlayerCorrect />;
+      return <PlayerCorrect pts={scoreDiff} />;
     } else if (playerState === "seeing-answer-incorrect") {
       bottomButtons = false;
-      return <PlayerIncorrect />;
+      return <PlayerIncorrect consolationPts={scoreDiff} />;
```
Comment (josephmaxwellhyland) [inline front-end/src/player/PlayerApp.tsx:125]: this score diff will work only if the player does not refresh. otherwise, it will simply not show a score.

**Review by hylandtechoutreach — APPROVED** [review]

**Pull request merged**